Add kafka source

vrazdalovschi opened the pull request:
Closes #9

snowplowcla commented:
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

vrazdalovschi commented:
[colmsnowplow](https://github.com/colmsnowplow), Let me know your thoughts about this PR. The code is based on the `feat/kafka_source` branch. I have added additional e2e tests and additional configuration parameters, fixed minor things. I'd be happy to move forward with all comments and progress on making it live ASAP.

vrazdalovschi commented:
I signed it!

snowplowcla commented:
Confirmed! @vrazdalovschi has signed the Contributor License Agreement. Thanks so much.

colmsnowplow commented:
Thanks @vrazdalovschi! 

It's turning out to be quite a busy week so please bear with me, I will take a look as soon as I can. :)

colmsnowplow commented on pkg/source/kafka/kafka_source.go:
```diff
@@ -0,0 +1,361 @@
+	SASLUsername     string `hcl:"sasl_username,optional" env:"SOURCE_KAFKA_SASL_USERNAME" `
+	SASLPassword     string `hcl:"sasl_password,optional" env:"SOURCE_KAFKA_SASL_PASSWORD"`
+	SASLAlgorithm    string `hcl:"sasl_algorithm,optional" env:"SOURCE_KAFKA_SASL_ALGORITHM"`
+	TLSCert          string `hcl:"tls_cert,optional" env:"SOURCE_KAFKA_TLS_CERT_B64"`
+	TLSKey           string `hcl:"tls_key,optional" env:"SOURCE_KAFKA_TLS_KEY_B64"`
+	TLSCa            string `hcl:"tls_ca,optional" env:"SOURCE_KAFKA_TLS_CA_B64"`
```
nit: We should name these the same way we do for the target: https://github.com/snowplow/snowbridge/blob/623842cbf785ea30ab71fa974b4cf58cd42e299a/pkg/target/kafka.go#L42-L44

(they expect filepaths rather than a previous iteration which was b64)

colmsnowplow commented on pkg/source/kafka/kafka_source.go:
```diff
@@ -0,0 +1,361 @@
+			err := consumer.source.WriteToTarget(messages)
+			if err != nil {
+				consumer.log.Debugf("Error writing to target: %s", err)
+				consumeErr = err
```
I think this should just be:

`consumer.log.WithFields(log.Fields{"error": err}).Error(err)`

When `WriteTotarget` returns an error it just means we failed to send some data - these messages won't have been acked so they'll get retried eventually. When we return it via `consumeErr`, it'll cause the app to crash and reboot, which is not the desired behaviour here.

colmsnowplow commented on pkg/source/kafka/kafka_source.go:
```diff
@@ -0,0 +1,361 @@
+		log:              ks.log,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
```
I think it's better to use `cctx, cancel := context.WithCancel(ctx)` here like we do in the pubsub source

colmsnowplow replied:
Sorry, I left this comment thinking that elsewhere in the project I had been creating new contexts rather than using background. When I look at it again, there's no difference between what I asked for and what you had. I didn't intend to ask you to do pointless renaming, my apologies 🤦

colmsnowplow commented on pkg/source/kafka/kafka_source.go:
```diff
@@ -0,0 +1,361 @@
+		if err := ks.client.Consume(ctx, strings.Split(ks.topic, ","), &consumer); err != nil {
+			return err
+		}
+		if ctxErr := ctx.Err(); ctxErr != nil {
```
I think I'd rather we don't ignore this, in case the context gets cancelled when we don't expect it to be. It's not a big deal if the error gets reported on exit I think.

colmsnowplow commented on pkg/source/kafka/kafka_source.go:
```diff
@@ -0,0 +1,361 @@
+	sarama.Logger = logger
+
+	saramaConfig := sarama.NewConfig()
+	saramaConfig.ClientID = "snowplow_stream_replicator"
```
nit

```suggestion
	saramaConfig.ClientID = "snowplow_snowbridge"
```

colmsnowplow commented on pkg/source/kafka/kafka_source.go:
```diff
@@ -0,0 +1,361 @@
+	return fmt.Sprintf("brokers:%s:topic:%s", ks.brokers, ks.topic)
+}
+
+func getKafkaVersion(targetVersion string) (sarama.KafkaVersion, error) {
```
This is already defined in the target, we can factor it out into `pkg/common/helpers.go`

colmsnowplow commented on pkg/source/kafka/kafka_source.go:
```diff
@@ -0,0 +1,361 @@
+				break
+			}
+		}
+		if !supportedVersion {
+			return sarama.DefaultVersion, fmt.Errorf("unsupported version `%s`. select older, compatible version instead", parsedVersion)
+		}
+	}
+
+	return preferredVersion, nil
+}
+
+// SHA256 hash
+var SHA256 scram.HashGeneratorFcn = func() hash.Hash { return sha256.New() }
+
+// SHA512 hash
+var SHA512 scram.HashGeneratorFcn = func() hash.Hash { return sha512.New() }
+
+type xdgSCRAMClient struct {
+	*scram.Client
+	*scram.ClientConversation
+	scram.HashGeneratorFcn
+}
+
+func (x *xdgSCRAMClient) Begin(userName, password, authzID string) (err error) {
+	x.Client, err = x.HashGeneratorFcn.NewClient(userName, password, authzID)
+	if err != nil {
+		return err
+	}
+	x.ClientConversation = x.Client.NewConversation()
+	return nil
+}
+
+func (x *xdgSCRAMClient) Step(challenge string) (response string, err error) {
+	response, err = x.ClientConversation.Step(challenge)
+	return
+}
+
+func (x *xdgSCRAMClient) Done() bool {
+	return x.ClientConversation.Done()
+}
```
All of this is already defined in the target, we can factor it out into `pkg/common/helpers.go`

colmsnowplow reviewed (commented):
Hi @vrazdalovschi,

Apologies for the delay - along with a hectic schedule I had to take some unplanned leave.

Thanks for bringing the old PR up to the date with the latest state of the repo.

Review is just based on a reading of the code for now but largely looks good, just a few comments. 

---

The old PR never got merged because the tests didn't demonstrate the desired behaviour:

The `testutil.ReadAndReturnMessages()` (in combination with `testutil.DefaultTestWriteBuilder()`) function should receive exactly the amount of data that the source provides, and then return those messages after there are no more messages for the provided timeout period. So, for example, the idea is that given this usage:

```
testutil.ReadAndReturnMessages(s, 3*time.Second, testutil.DefaultTestWriteBuilder, nil)
```

The test would receive any data that comes from s, and then return the data in a slice of models.Message after 3 seconds.

Looking at this PR, I strongly suspect that the reason for our issues previously was in how the mocking is done. I think probably the easiest way to get past it & provide sufficient is to just dispense with the mocks for now, and do it via an integration test.

I think we can get the testing we need fairly easily by just copying [the parallel test for pubsub source](https://github.com/snowplow/snowbridge/blob/master/pkg/source/pubsub/pubsub_source_test.go#L32), swapping out the pubsub topic etc., creation for the same thing that's done for kafka topics in the e2e tests.

If you're able to look into that then great, otherwise it'll have to wait for someone on our side to find time to put to that task. :)

vrazdalovschi commented:
@colmsnowplow, Hi. I have realized that I do not need the SnowBridge project in my stack right now, but I'd like to finish this PR, so there is no pushing to the timeline. Right now, I'm pretty busy with other personal things for the next ~10 days, but I'll try in the meantime to apply most of the fixes and take a look at the testing. If the test is time-consuming, then I might delay that part, and anyone could jump here :) 
Thank you for your review!

colmsnowplow commented:
No problem, no rush. When you do get to it happy to discuss/offer whatever help I can. Hopefully we find some space to chip away at it our side too :)

vrazdalovschi commented:
@colmsnowplow , hopefully, I have found some time and fixed the notes. The one area for improvement for me is testing. The e2e test is present in `release_test/e2e_source_test.go`. I haven't added mocked tests "integration" tests for kafka source. In case you want to include them, could you describe them?
1. To test happy flow, I have to mock the Consume func and to get that printed in STDOUT? 
2. Anything else?

colmsnowplow commented:
I might get to the in-code comments today but in case I don't, addressing the main Q:

> I haven't added mocked tests "integration" tests for kafka source. In case you want to include them, could you describe them?

> To test happy flow, I have to mock the Consume func and to get that printed in STDOUT?

"Integration test" normally describes something a bit different, so perhaps wrong word used. I just mean that we can use the existing local kafka cluster setup to "cheat" on unit tests and use those resources instead of mocks, to isolate the source function behaviour in a test.

Effectively this is covered in the end to end tests, but it's important to have because we need a way to easily isolate/replicate scenarios when we need to debug stuff etc. I'm especially cautious in this case since the original PR never reached a point where it proved itself to work via unit tests (so hesitant to merge to prod without a means of prodding at it basically).

The way this works is that when running tests locally, we can run with `-short` flag to skip the tests that rely on local resources like I propose here, or if we want to run those, we run `make integration-up` to set those up first. There's a kafka instance in the docker compose file that that command runs, so we'd basically be using that instead of mocking stuff.

The below is untested, but I've taken the parallel tests from pubsub source and will pseudocode out what I'd do now:

```
func TestKafkaSource_ReadAndReturnSuccessIntegration(t *testing.T) {
	if testing.Short() {
		t.Skip("skipping integration test")
	}
	assert := assert.New(t)

	// Create topic and subscription
	// topic, subscription := testutil.CreatePubSubTopicAndSubscription(t, "test-topic", "test-sub")
	// defer topic.Delete(context.Background())
	// defer subscription.Delete(context.Background())

// ^^ replace with what you've already done here: https://github.com/snowplow/snowbridge/blob/8dd6f57fbead54f194e054d29aa6e9550cc8eebe/release_test/e2e_source_test.go#L162C2-L178

	// Write to topic
	// testutil.WriteToPubSubTopic(t, topic, 10)

// ^^ replace with some code to write some messages to the kafka topic

// Then, below, just change the config stuff and the vars to be kafka source stuff, and it should be done.
	t.Setenv("SOURCE_NAME", "pubsub")
	t.Setenv("SOURCE_PUBSUB_SUBSCRIPTION_ID", "test-sub")
	t.Setenv("SOURCE_PUBSUB_PROJECT_ID", `project-test`)

	adaptedHandle := adapterGenerator(configFunction)

	pubsubSourceConfigPair := config.ConfigurationPair{Name: "pubsub", Handle: adaptedHandle}
	supportedSources := []config.ConfigurationPair{pubsubSourceConfigPair}

	pubsubConfig, err := config.NewConfig()
	assert.NotNil(pubsubConfig)
	assert.Nil(err)

	pubsubSource, err := sourceconfig.GetSource(pubsubConfig, supportedSources)

	assert.NotNil(pubsubSource)
	assert.Nil(err)
	assert.Equal("projects/project-test/subscriptions/test-sub", pubsubSource.GetID())

	output := testutil.ReadAndReturnMessages(pubsubSource, 5*time.Second, testutil.DefaultTestWriteBuilder, nil)
	assert.Equal(10, len(output))
	for _, message := range output {
		assert.Contains(string(message.Data), `message #`)
		assert.Nil(message.GetError())
	}
}
```

Like I mentioned - it's totally fine if you don't do this, you've already added the e2e stuff which is great. I'll just need to find space to add this kind of test (and then think about other test cases) before we merge. :)

colmsnowplow commented:
Released in https://github.com/snowplow/snowbridge/pull/285 - thanks again @vrazdalovschi
